src/main/java/org/eclipse/uprotocol/mqtt/HiveMqMQTT5Client.java

```diff
@@ -0,0 +1,280 @@
+        LOG.trace("should send a message:\n{}", uMessage);
+        CompletableFuture<UStatus> result = new CompletableFuture<>();
+
+        Mqtt5UserProperties userProperties = Mqtt5UserProperties.builder()
+                .add("0", "1")
+                .add(USER_PROPERTIES_KEY_FOR_ID, UuidSerializer.serialize(uMessage.getAttributes().getId()))
+                .add(USER_PROPERTIES_KEY_FOR_MESSAGE_TYPE, Integer.toString(uMessage.getAttributes().getTypeValue()))
+                .add(USER_PROPERTIES_KEY_FOR_SOURCE_NAME, uMessage.getAttributes().getSource().getAuthorityName())
+                .add(USER_PROPERTIES_KEY_FOR_SINK_NAME, uMessage.getAttributes().getSink().getAuthorityName())
+                .add(USER_PROPERTIES_KEY_FOR_PRIORITY, Integer.toString(uMessage.getAttributes().getPriorityValue()))
+                .add(USER_PROPERTIES_KEY_FOR_TTL, Integer.toString(uMessage.getAttributes().getTtl()))
+                .add(USER_PROPERTIES_KEY_FOR_PERMISSION_LEVEL, Integer.toString(uMessage.getAttributes().getPermissionLevel()))
+                .add(USER_PROPERTIES_KEY_FOR_COMMSTATUS, Integer.toString(uMessage.getAttributes().getCommstatusValue()))
+                .add(USER_PROPERTIES_KEY_FOR_REQID, UuidSerializer.serialize(uMessage.getAttributes().getReqid()))
+                .add(USER_PROPERTIES_KEY_FOR_TOKEN, uMessage.getAttributes().getToken())
+                .add(USER_PROPERTIES_KEY_FOR_TRACEPARENT, uMessage.getAttributes().getTraceparent())
+                .add(USER_PROPERTIES_KEY_FOR_PAYLOAD_FORMAT, Integer.toString(uMessage.getAttributes().getPayloadFormatValue()))
+                .build();
+
+        Mqtt5PublishBuilder.Send.Complete<CompletableFuture<Mqtt5PublishResult>> sendHandle = client.publishWith()
+                .topic(getTopicForSending(uMessage.getAttributes().getSource(), uMessage.getAttributes().getSink()))
+                .payload(uMessage.getPayload().toByteArray())
+                .userProperties(userProperties);
+
+        if (uMessage.getAttributes().hasTtl()) {
+            sendHandle.messageExpiryInterval(Math.round(uMessage.getAttributes().getTtl() / 1000d));
+        }
+
+        sendHandle
+                .send()
+                .whenCompleteAsync((mqtt5PublishResult, throwable) -> {
+                    LOG.trace("got complete callback");
+                    if (throwable != null) {
+                        LOG.error("Error sending message", throwable);
+                        result.complete(UStatus.newBuilder().setCode(INTERNAL).build());
+                    } else {
+                        LOG.trace("publishResult sending message {}", mqtt5PublishResult);
+                        result.complete(UStatus.newBuilder().setCode(OK).build());
+                    }
+                });
+
+
+        return result;
+    }
+
+    public CompletableFuture<UStatus> registerListenerAsync(UUri sourceFilter, UUri sinkFilter, UListener listener) {
```

Comment:
These can be private, only used by the sync calls.

Reply:
as the interface methods are now async, I just inlined the private methods in the interface methods